### DISCUSSION TITLE
[Outdated] Benchmarks

### DISCUSSION THREAD

**Discussion by awf** [Mikolaj/mostly-harmless#2] [General]
Testing this as a place where benchmark results might be pasted.   I'll try to do an initial example...

**Comment by awf** [comment]
Model: prod
Setting: model only, list of doubles
Size: 5e8
Time: 1.04 +- .21/48% sec

```Haskell
prod :: Fractional a => [a] -> a
prod [x] = x
prod (x:xs) = x * prod xs
```

```
benchmarking 5e8/func
time                 1.058 s    (691.1 ms .. 1.778 s)
                     0.944 R²   (0.915 R² .. 1.000 R²)
mean                 1.037 s    (821.8 ms .. 1.164 s)
std dev              213.6 ms   (79.58 ms .. 293.8 ms)
variance introduced by outliers: 48% (moderately inflated)
```

**Comment by Mikolaj** [comment]
For a simple example, I'm including the original email with prod benchmarks, but with all numbers updated, as of HEAD of master branch right now, GHC 8.10.7 (backprop and deps haven't updated yet). Also, the 5e8 nonsense is corrected (it was meant to be 0.5 * 10^8). In particular, computing the normal value (vec_func) of a function (expressed in our formalism for computing gradients) is now 40 times faster than in the email (that's still much slower than without the overhead of our formalism; no benchmark for that exact comparison is present, though, because no other approach uses unboxed vectors).

I mark the text down slightly, but otherwise leave it unchanged. An obvious improvement would be to cite code as a github snippet that links to the codebase in the repo (and probably offers some other benefits; I'm low-tech, I don't remember). [Edit: I've now added this for our and the ad library. Is that better?] I see my machine was bit busier, so most runtimes are slightly higher than in the email, but I hope it's mostly consistent.

A criterion html report rougly corresponding to this data (probably a bit more accurate) is committed to the repo and viewable at
https://combinatronics.com/Mikolaj/mostly-harmless/b9f1ed13b7e549f7f221f2ea4a1033ce59677247/misc-bench/ProdAll-78ba46b.html

--

> And the other example I always enjoy is simply the product of all vector elements (sometimes called “Speelpenning’s example”), which has various hand-written gradient implementations at https://github.com/microsoft/knossos-ksc/blob/master/users/awf/prod.hs

This is indeed a great benchmark --- not completely artificial
(though we are probably firmly stuck at NaN after multiplying
just a few initial millions of random numbers) and it lets us
get to almost 10^8 parameters (precisely to 0.5 * 10^8, because
at 10^8 all but the baseline contenders run out of memory).
A lot of the time is spent in GC that doesn't fit in cache
(up to 40G is taken, depending on benchmark and contender),
hence the large variance, but results are mostly repeatable.
Tweaking GC knobs in GHC RTS might change a lot, but I kept
them all at default.

This benchmarking resulted in a couple of optimizations
to our library that also seem to help with allocations and runtime
of the neural networks examples.

First, the baseline, the manually coded gradient from the file
at your link, applied to a list of 0.5 * 10^8 random Doubles.

```Haskell
prod :: Fractional a => [a] -> a
prod [x] = x
prod (x:xs) = x * prod xs
```

```
benchmarking Half1e8/func   -- 5e7 == 5 * 1e7 == 5 * 10^7 == 0.5 * 10^8 == 0.5 * 1e8 == 0.5e8
time                 1.134 s    (718.6 ms .. 1.870 s)
                     0.946 R²   (0.922 R² .. 1.000 R²)
mean                 1.111 s    (887.0 ms .. 1.271 s)
std dev              220.9 ms   (87.27 ms .. 302.0 ms)
variance introduced by outliers: 48% (moderately inflated)
```

And the manually coded fast gradient applied to the Doubles.

```Haskell
grad_prod_aux :: Fractional a => a -> [a] -> (a, [a])
grad_prod_aux _ [] = (1.0, [])
grad_prod_aux q (x:xs) =
    let (p1,out) = grad_prod_aux (q * x) xs
    in (x * p1, q * p1 : out)

grad_prod :: Fractional a => [a] -> [a]
grad_prod xs = snd $ grad_prod_aux 1.0 xs
```

```
benchmarking Half1e8/grad
time                 13.09 s    (11.68 s .. 14.17 s)
                     0.998 R²   (0.998 R² .. 1.000 R²)
mean                 13.43 s    (13.15 s .. 13.66 s)
std dev              281.0 ms   (184.3 ms .. 351.9 ms)
variance introduced by outliers: 19% (moderately inflated)
```

Now an alternative baseline, that is, a different handwritten
gradient, from the backprop library. I have no idea what's
the numerical quality of this one.

```Haskell
handwritten_vec_prod_aux :: (Fractional r, Backprop r, Reifies s W)
                         => BVar s (Data.Vector.Vector r) -> BVar s r
handwritten_vec_prod_aux = Prelude.Backprop.product  -- hand-written gradient
```
{-
```Haskell
product :: (Foldable t, Functor t, Backprop (t a), Fractional a, Reifies s W)
        => BVar s (t a) -> BVar s a
product af = liftOp1 af . op1 $ \xs ->
    let p = P.product xs
    in ( p
       , \d -> (\x -> p * d / x) P.<$> xs
       )
```
-}
```Haskell
handwritten_vec_prod :: Data.Vector.Vector Double -> Double
handwritten_vec_prod = evalBP handwritten_vec_prod_aux
```

```
benchmarking Half1e8/handwritten_vec_func
time                 402.6 ms   (393.7 ms .. 412.1 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 398.4 ms   (395.4 ms .. 400.1 ms)
std dev              2.866 ms   (1.008 ms .. 3.837 ms)
variance introduced by outliers: 19% (moderately inflated)
```

```Haskell
handwritten_vec_grad_prod :: Data.Vector.Vector Double
                          -> Data.Vector.Vector Double
handwritten_vec_grad_prod = gradBP handwritten_vec_prod_aux
```

```
benchmarking Half1e8/handwritten_vec_grad
time                 5.650 s    (4.840 s .. 6.836 s)
                     0.995 R²   (0.986 R² .. NaN R²)
mean                 5.644 s    (5.351 s .. 5.852 s)
std dev              321.9 ms   (179.6 ms .. 454.7 ms)
variance introduced by outliers: 19% (moderately inflated)
```

And finally automatic differentiation results. Backprop dropped off,
because it takes forever even for 10^5 parameters. I suppose,
it's tailored to working with huge matrices from hmatrix,
with gradients of matrix operations written by hand, instead
of derived from gradients of their arithmetic constituents.

Results with our library.

https://github.com/Mikolaj/mostly-harmless/blob/99d1d85d16cf4dbf95b5b11ec0b547ea51d3391f/bench/ProdMostlyHarmlessTools.hs#L110-L118

```
benchmarking Half1e8/vec_func
time                 240.0 ms   (237.7 ms .. 242.2 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 239.3 ms   (238.7 ms .. 240.0 ms)
std dev              942.8 μs   (532.9 μs .. 1.331 ms)
variance introduced by outliers: 16% (moderately inflated)
```

https://github.com/Mikolaj/mostly-harmless/blob/99d1d85d16cf4dbf95b5b11ec0b547ea51d3391f/bench/ProdMostlyHarmlessTools.hs#L120-L122

```
benchmarking Half1e8/vec_grad
time                 12.14 s    (8.171 s .. 16.34 s)
                     0.985 R²   (0.947 R² .. 1.000 R²)
mean                 13.20 s    (12.10 s .. 14.58 s)
std dev              1.509 s    (429.1 ms .. 2.044 s)
variance introduced by outliers: 23% (moderately inflated)

```

Results with ad with 'ffi' flag:

https://github.com/Mikolaj/mostly-harmless/blob/99d1d85d16cf4dbf95b5b11ec0b547ea51d3391f/ad-bench/ProdAdTools.hs#L95-L97

```
benchmarking Half1e8/func
time                 1.055 s    (697.4 ms .. 1.792 s)
                     0.942 R²   (0.912 R² .. 1.000 R²)
mean                 1.036 s    (819.1 ms .. 1.191 s)
std dev              216.4 ms   (83.91 ms .. 298.6 ms)
variance introduced by outliers: 48% (moderately inflated)
```

https://github.com/Mikolaj/mostly-harmless/blob/99d1d85d16cf4dbf95b5b11ec0b547ea51d3391f/ad-bench/ProdAdTools.hs#L99-L100

```
benchmarking Half1e8/grad
time                 34.01 s    (33.67 s .. 34.44 s)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 36.10 s    (35.21 s .. 37.53 s)
std dev              1.367 s    (246.4 ms .. 1.780 s)
variance introduced by outliers: 19% (moderately inflated)
```

Please remember the saying:
Lies, damned lies, and benchmarks.

To make this more trustworthy, we'd need testsuites based
on the same code that guarantee results are not buggy and
decently precise and also a tweak to keep us clear of NaNs.

**Comment by Mikolaj** [comment]
A bonus entry, library ad in default mode (without the `ffi` flag when building, but exactly the same code as above):

```
benchmarking Half1e8/func
time                 1.121 s    (720.3 ms .. 1.875 s)
                     0.944 R²   (0.917 R² .. 1.000 R²)
mean                 1.109 s    (888.3 ms .. 1.269 s)
std dev              220.1 ms   (82.75 ms .. 302.4 ms)
variance introduced by outliers: 48% (moderately inflated)

benchmarking Half1e8/grad
time                 42.83 s    (40.33 s .. 45.69 s)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 44.19 s    (43.50 s .. 44.88 s)
std dev              819.9 ms   (542.6 ms .. 963.9 ms)
variance introduced by outliers: 19% (moderately inflated)
```

**Comment by Mikolaj** [comment]
[Edit: I initially conflated this with selective fusion, while the omitting of Delta expression bindings only exposes fusion opportunities (with no bindings splitting the expression into tiny parts, the whole expression can be simplified more easily, sometimes even automatically). Fusion probably wouldn't help in this particular case, anyway.]

Second bonus: prod, run with our library, but with manually omitting Delta expression bindings (but still automatic gradient, derived from the gradient we defined for a binary multiplication). Only one let binding, at the top, as opposed to, previously, one binding for each binary multiplication. 

https://github.com/Mikolaj/mostly-harmless/blob/99d1d85d16cf4dbf95b5b11ec0b547ea51d3391f/bench/ProdMostlyHarmlessTools.hs#L134-L142

```
benchmarking Half1e8/omit_vec_func
time                 238.8 ms   (235.7 ms .. 244.4 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 236.9 ms   (235.7 ms .. 238.1 ms)
std dev              1.546 ms   (1.243 ms .. 1.661 ms)
variance introduced by outliers: 16% (moderately inflated)
```

https://github.com/Mikolaj/mostly-harmless/blob/99d1d85d16cf4dbf95b5b11ec0b547ea51d3391f/bench/ProdMostlyHarmlessTools.hs#L144-L146

```
benchmarking Half1e8/omit_vec_grad
time                 11.10 s    (8.262 s .. 14.80 s)
                     0.988 R²   (0.961 R² .. 1.000 R²)
mean                 12.25 s    (11.32 s .. 13.37 s)
std dev              1.270 s    (416.8 ms .. 1.653 s)
variance introduced by outliers: 23% (moderately inflated)
```

That the runtime of computing a value is unchanged is not surprising --- the total Delta expression is smaller (all `Var` constructors gone), but it's computed in one go, spanning beyond CPU cache. OTOH with many let-bindings, it's computed and discarded in tiny chunks by our dummy monad for computing function values. 

However, I'm surprised the runtime of gradient computation is so close to the original. Perhaps ~selective fusion~ omitting Delta expression bindings is not so important, after all? We'd better measure this in less artificial programs.

**Reply by awf** [reply]
> Second bonus: prod, run with our library, but with manually omitting Delta expression bindings (but still automatic gradient, derived from the gradient we defined for a binary multiplication). Only one let binding, at the top, as opposed to, previously, one binding for each binary multiplication.

Would you mind linking to code for this?  I'm fairly sure I could find it, but I would fear making a mistake and looking at the wrong thing.

**Reply by Mikolaj** [reply]
Yes, of course. Here you are.

**Reply by awf** [reply]
Fab, much more reassuring, thanks!

**Comment by awf** [comment]
BTW, regarding NaNs in prod, if we initialise with numbers "around 1", e.g. exponential of normally distributed randoms, that should reduce the incidence of NaN?

**Reply by awf** [reply]
Ah, I see you are already doing something similar
https://github.com/Mikolaj/mostly-harmless/blob/1404371c24a0f173cacebe4e7ce747d5c7bda260/ad-bench/ProdAd.hs#L12

It might still be worth trying exp of -0.5..0.5

**Reply by Mikolaj** [reply]
Oh, I just copied this from the file you have sent me. This one: https://github.com/Mikolaj/mostly-harmless/blob/master/CREDITS#L17

Current code should give numbers between 0.55 and 1.55, because `randoms` is between 0 and 1. And I can see in `ghci`:

```
Prelude> exp (-0.5)
0.6065306597126334
Prelude> exp 0.5
1.6487212707001282
```

It's possible to explicitly request just the numbers between these bounds. Is that what you are suggesting? Or perhaps it already works well, given how close the two bound sets are and I just panicked needlessly.

No problem if NaN occurs sometimes --- we just need to check it doesn't with the seed we use for the benchmark, in case CPU heats less with NaNs, but only when it's added from the left, not from the right, and the tested libraries differ exactly in that. Also, tests, when we finally set them up, require few NaNs.

**Comment by Mikolaj** [comment]
Newest benchmarks are linked in https://github.com/Mikolaj/mostly-harmless/blob/master/README.md
